File: CodeForge/Sources/CreedFlow/Services/ProcessTracker.swift
import Foundation

/// Global registry of child processes spawned by CLI backends.
/// Ensures all child processes are terminated when the app exits.
package final class ProcessTracker: @unchecked Sendable {
    package static let shared = ProcessTracker()

    private let lock = NSLock()
    private var processes: Set<Int32> = [] // tracked by PID

    private init() {}

    /// Register a running process for cleanup on app termination.
    package func track(_ process: Process) {
        guard process.isRunning else { return }
        lock.lock()
        processes.insert(process.processIdentifier)
        lock.unlock()
    }

    /// Unregister a process (already exited normally).
    package func untrack(_ process: Process) {
        lock.lock()
        processes.remove(process.processIdentifier)
        lock.unlock()
    }

    /// Terminate all tracked child processes. Called from applicationWillTerminate.
    package func terminateAll() {
        lock.lock()
        for pid in processes {
            kill(pid, SIGTERM)
        }
        processes.removeAll()
        lock.unlock()
    }
}

File: creedflow-desktop/src-tauri/src/services/process_tracker.rs
use std::collections::HashSet;
use std::sync::Mutex;

/// Global singleton tracking all spawned CLI child processes.
/// On app exit, all children receive termination signals — no orphans.
pub static PROCESS_TRACKER: ProcessTracker = ProcessTracker::new();

pub struct ProcessTracker {
    pids: Mutex<HashSet<u32>>,
}

impl ProcessTracker {
    const fn new() -> Self {
        Self {
            pids: Mutex::new(HashSet::new()),
        }
    }

    pub fn track(&self, pid: u32) {
        if pid == 0 { return; }
        self.pids.lock().unwrap().insert(pid);
    }

    pub fn untrack(&self, pid: u32) {
        self.pids.lock().unwrap().remove(&pid);
    }

    pub fn terminate_all(&self) {
        let pids: Vec<u32> = self.pids.lock().unwrap().drain().collect();
        for pid in pids {
            terminate_process(pid);
        }
    }
}

// Const Mutex::new is unstable — use lazy_static alternative
// Actually in Rust 1.63+ Mutex::new is const, so this works.

#[cfg(unix)]
pub fn terminate_process(pid: u32) {
    unsafe {
        libc::kill(pid as i32, libc::SIGTERM);
    }
}

#[cfg(windows)]
pub fn terminate_process(pid: u32) {
    use windows::Win32::System::Threading::{OpenProcess, TerminateProcess, PROCESS_TERMINATE};
    use windows::Win32::Foundation::CloseHandle;
    unsafe {
        if let Ok(handle) = OpenProcess(PROCESS_TERMINATE, false, pid) {
            let _ = TerminateProcess(handle, 1);
            let _ = CloseHandle(handle);
        }
    }
}

#[cfg(not(any(unix, windows)))]
pub fn terminate_process(_pid: u32) {
    log::warn!("Process termination not supported on this platform");
}
